### PR TITLE
Support Godot 4 w/ 3.x compatibility

### PR DESCRIPTION
**Pull request by RGonzalezTech**
## Summary

I added another tester godot project that uses Godot 4.0.3 and GUT 9.0.1 & I renamed the existing tester to `tester_GUT_v7.4.1` so that we can open the right project depending on the Godot version.

I refactored the code to make heavy use of functions so that we can easily locate & isolate the behavior that needs version-handling logic.

The download URL and headless behavior changed in Godot v4. We no longer need to download a binary compiled to run headlessly, we simply append `--headless` to the executable args.

Also, the architecture suffix changed from `_64` to `_x86_64` or `.64`  to `.x86_64`

Luckily, the GUT command line interface and results XML did not change.

## Details

I used the following function to determine whether this version is greater than v3:

```bash
check_godot_version() {
    local version=$GODOT_VERSION
    local version_three=3

    # Remove the leading 'v' character if present
    version=${version#v}

    # Extract the major version number from the version string
    local current_version=$(echo "$version" | cut -d. -f1)

    if [ "$current_version" -gt "$version_three" ]; then
        # v4+ behavior
        IS_VERSION_FOUR=1
    fi
}
```

It updates a global "boolean" called `IS_VERSION_FOUR`.

### Downloading Godot

The download filename changed slightly. That URL is generated in the `generate_dl_url` function. You can see the change below:

```bash
generate_dl_url_three() {
    echo "generating download url ...v3"

    if [ "$IS_MONO" = "true" ]; then
        FULL_GODOT_NAME="${FULL_GODOT_NAME}_mono_linux_headless"
        FULL_GODOT_NAME_EXT="${FULL_GODOT_NAME}_64"
    else
        FULL_GODOT_NAME="${FULL_GODOT_NAME}_linux_headless"
        FULL_GODOT_NAME_EXT="${FULL_GODOT_NAME}.64"
    fi
}

generate_dl_url_four() {
    echo "generating download url ...v4"

    if [ "$IS_MONO" = "true" ]; then
        FULL_GODOT_NAME="${FULL_GODOT_NAME}_mono_linux"
        FULL_GODOT_NAME_EXT="${FULL_GODOT_NAME}_x86_64"
    else
        FULL_GODOT_NAME="${FULL_GODOT_NAME}_linux"
        FULL_GODOT_NAME_EXT="${FULL_GODOT_NAME}.x86_64"
    fi
}

generate_dl_url() {
    if [ "$RELEASE_TYPE" = "stable" ]; then
        GODOT_URL_PATH_SUBDIR=""
    else
        # if not stable, then it's a release candidate or beta
        # Those are located in a subdirectory of the release type
        # example: /beta1, /rc1, etc.
        GODOT_URL_PATH_SUBDIR="/${RELEASE_TYPE}"
    fi
    # This is the path to the godot version hosted on tuxfamily
    # example: 3.2.3/, 3.2.3/beta1/, etc.
    GODOT_URL_PATH="${GODOT_VERSION}${GODOT_URL_PATH_SUBDIR}/"

    if [ "$IS_MONO" = "true" ]; then
        # mono builds are in a subdirectory of the godot version
        # example: 3.2.3/mono/, 3.2.3/beta1/mono/, etc.
        GODOT_URL_PATH="${GODOT_URL_PATH}mono/"
    fi

    FULL_GODOT_NAME="Godot_v${GODOT_VERSION}-${RELEASE_TYPE}"
    # Different behavior for v4+ and v3
    if [[ "$IS_VERSION_FOUR" -eq "1" ]]; then
        generate_dl_url_four
    else
        generate_dl_url_three
    fi

    # Apply the generated path & name to the download url
    DL_URL="https://downloads.tuxfamily.org/godotengine/${GODOT_URL_PATH}${FULL_GODOT_NAME_EXT}.zip"

    if [ "$CUSTOM_GODOT_DL_URL" != "" ]; then
        DL_URL="$CUSTOM_GODOT_DL_URL"
    fi
}
```

### Generate Run Options

Godot v4 no longer has a headless build, you can simply add the `--headless` argument to the regular linux executable and it will run headlessly. This is accounted for in the `generate_run_options` function, where we append `--headless` to the run options if we're on v4+

```bash
    if [ "$IS_VERSION_FOUR" -eq "1" ]; then
        # v4+ behavior
        # you must use the --headless argument 
        # rather than use a headless binary
        RUN_OPTIONS="${RUN_OPTIONS} --headless"
    fi
```

### Test Running & Analysis

While the code has been refactored to make heavier use of functions, the actual behavior hasn't changed.

### Version Bump

I changed the README a bit with more relevant information and bumped the action to v4

## Relevant issues

#26 #25

### CODE REVIEW

**Comment by RGonzalezTech** [comment]
I'm going to sleep for the night 😴 I'll review tomorrow morning

**Comment by croconut** [comment]
@RGonzalezTech let's update the actions to run the test workflow on PR / commit to PR

File: test_runner_lib/functions.sh [hunk 1]
```diff
@@ -0,0 +1,253 @@
+# credit: https://stackoverflow.com/questions/24283097/reusing-output-from-last-command-in-bash
+# capture the output of a command so it can be retrieved with ret
+cap() { tee /tmp/capture.out; }
+
+# return the output of the most recent command that was captured by cap
+ret() { cat /tmp/capture.out; }
+
+# Parse XML with bash-only: 
+# https://stackoverflow.com/questions/893585/how-to-parse-xml-in-bash/2608159#2608159
+rdom () { local IFS=\> ; read -d \< E C ;}
```
Comment (baabcantcode) [inline test_runner_lib/functions.sh:10]: just wondering why this is named rdom? is that some kind of xml thing im too json to understand? lol

File: .github/workflows/tester.yml [hunk 2]
```diff
@@ -7,11 +7,51 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
+    - uses: ./
```
Comment (baabcantcode) [inline .github/workflows/tester.yml:10]: please convert this to a branch on the repo and replace the current action `on` designation with 
`on: [pull_request]` that way we'll have records of the test successes on this repo

**Review by baabcantcode — DISMISSED** [review]
the code LGTM, just some ops based changes requested.

File: tester_GUT_v9.0.1/.gitignore [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.vscode/
+# Godot-specific ignores
+.import/
+
+export.cfg
+
+#really need to make sure this has no keys in it, btw
+#export_presets.cfg
```
[the remark this answers is not in the document]
Comment (RGonzalezTech) [reply tester_GUT_v9.0.1/.gitignore:9]: Handled in f8c973ed0e08b5f4b9ea7ca2b0884ff55c746366 and bbb5779574e3da6f665e0a9a1699bb800c6dc26e

File: test_runner_lib/functions.sh (hunk 1 above)
Comment (RGonzalezTech) [reply test_runner_lib/functions.sh]: Handled in 1ef4414ef8f6a338a1c698005250689129f1fe4e

File: .github/workflows/tester.yml (hunk 2 above)
Comment (RGonzalezTech) [reply .github/workflows/tester.yml:10]: Added PRs to the workflow trigger in 5608f92ff751d48ca773ce51d9703f4780872494

**Comment by RGonzalezTech** [comment]
## Summary

Refactored the code a bit. Due to the complexity of the download step & assert-check logic, I extracted those into functions and then I moved some of the less complex behaviors into functions as well to follow the [SLAP](https://medium.com/javarevisited/slap-that-ugly-code-6ec276d3a4bc) principle.

I understand the point, though. I noticed that I wasn't following the principle in the `generate_dl_url` function, so I flattened that one out in 5e70a57e3c06a85f015184e249da1ea89c670a21. I also cleaned up the `download_godot` and `run_tests` to better implement the SLAP principle.

I also removed some older behavior in 068fd8ed059477e2163f86fbca26225815f6a538.

Now, we're running into an issue with the imports.

## Importing Process Troubleshooting

Since I've removed the `.godot/` folder from `tester_GUT_v9.0.1`, the runs now fail. I've been trying to debug this in:

- d74dce223cec41e90e7ce930c239d8ff270861bd : added `--headless` to the import step
- 56de42d04992d04a2f755f50be8d710899f230c2 : tried removing the rebuilder scene

Not sure how to get it to rebuild the `.godot/` folder.

**Comment by RGonzalezTech** [comment]
Fixed the `.godot/` import issue. Had to increase the import timeout.

File: test_runner_lib/functions.sh [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        RUN_OPTIONS="${DIRECT_SCENE}"
+    elif [ "$CONFIG_FILE" != "res://.gutconfig.json" ]; then
+        RUN_OPTIONS="${RUN_OPTIONS} -gconfig=${CONFIG_FILE}"
+    fi
+
+    if [ "$IS_VERSION_FOUR" -eq "1" ]; then
+        # v4+ behavior
+        # you must use the --headless argument 
+        # rather than use a headless binary
+        RUN_OPTIONS="${RUN_OPTIONS} --headless"
```
Comment (croconut) [inline test_runner_lib/functions.sh:138]: Minor note, looks like this could be default too if it's just ignored in v3.

Comment (croconut) [reply test_runner_lib/functions.sh:138]: Looks like the biggest significant change for us with v4 is the download url name change and we can keep --headless as an always on argument.

**Review by croconut — APPROVED** [review]

**Comment by croconut** [comment]
@RGonzalezTech please merge + release when you're ready, LGTM :)

I'd recommend mentioning in release notes the deprecated argument

File: test_runner_lib/functions.sh (hunk 4 above)
Comment (RGonzalezTech) [reply test_runner_lib/functions.sh]: Looks like 3.x mono may not like `--headless` as an argument. Test run failed :( I'll revert and merge.

Comment (croconut) [reply test_runner_lib/functions.sh:138]: Sounds good, I think it's GUT that doesn't like it tho

Comment (croconut) [reply test_runner_lib/functions.sh:138]: Since we run the rebuilder scene with --headless I think

Comment (RGonzalezTech) [reply test_runner_lib/functions.sh:138]: That's fair. We can revisit it when we port to another language. Were you interested in porting this?

**Pull request merged**